trin-core/src/portalnet/overlay_service.rs

```diff
@@ -172,32 +239,138 @@ impl OverlayService {
         discovery: Arc<Discovery>,
         db: Arc<DB>,
         kbuckets: Arc<RwLock<KBucketsTable<NodeId, Node>>>,
+        bootnode_enrs: Vec<Enr>,
+        ping_queue_interval: Option<Duration>,
         data_radius: Arc<U256>,
         protocol: ProtocolId,
     ) -> Result<UnboundedSender<OverlayRequest>, String> {
         let (request_tx, request_rx) = mpsc::unbounded_channel();
+        let internal_request_tx = request_tx.clone();
+
         let overlay_protocol = protocol.clone();
 
+        let peers_to_ping = if let Some(interval) = ping_queue_interval {
+            HashDelayQueue::new(interval)
+        } else {
+            HashDelayQueue::default()
+        };
+
+        let (response_tx, response_rx) = mpsc::unbounded_channel();
+
         tokio::spawn(async move {
             let mut service = Self {
                 discovery,
                 db,
                 kbuckets,
                 data_radius,
                 protocol,
+                peers_to_ping,
                 request_rx,
+                request_tx: internal_request_tx,
+                active_requests: Arc::new(RwLock::new(HashMap::new())),
+                response_rx,
+                response_tx,
             };
 
-            info!("Starting {:?} overlay service", overlay_protocol);
+            // Attempt to insert bootnodes into the routing table in a disconnected state.
+            // If successful, then add the node to the ping queue. A subsequent successful ping
+            // will mark the node as connected.
+            //
+            // TODO: Perform FindNode lookup for local node's own node ID and refresh all kbuckets
+            // further away than closet neighbor (see Kademlia paper section 2.3).
+            for enr in bootnode_enrs {
+                let node_id = enr.node_id();
+                // TODO: Decide default data radius, and define a constant. Or if there is an
+                // associated database, then look for a radius value there.
+                let node = Node::new(enr, U256::from(u64::MAX));
+                let status = NodeStatus {
+                    state: ConnectionState::Disconnected,
+                    direction: ConnectionDirection::Outgoing,
```

Comment:
Yeah, I'm not entirely sure about this either, but my reasoning was that we are initiating the connection with some boot node(s) whose ENR(s) we know because that information was shared with us, not because that node connected to us.